### ISSUE TITLE
including tasks in dependent roles, when running the parent role via include_role

### ISSUE THREAD

**Issue by kraduk** [ansible/ansible#29541]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

include_role
##### ANSIBLE VERSION

```
# ansible --version
[DEPRECATED] [defaults]hostfile: The key is misleading as it can also be a list of hosts, a directory or a list of paths. It will be removed in 2.8. As alternative use one of [inventory]
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]

```

##### SUMMARY
Ok role A includes role B in its tasks. However Role B depends on role C. Role C has an include_tasks in its tasks. This all appears to work. However on close inspection it appears that the include_tasks in role C actually gets the file with a prefix of role Bs role_path. Therefore includes the wrong tasks.

eg

````
role A, tasks/main.yml

- name: setup jobs
    include_role:
      name: ROLEB

Role B, meta/main.yml

dependencies:
  - { role: ROLEC, when: full_deploy, var1: "{{var1}}", var2: "{{var2}}" }

Role C, tasks/main.yml

- include_tasks: tasks.yml


result

13:46:11 TASK [RoleC : include_tasks] ***********************************************
13:46:11 Monday 11 September 2017  13:46:11 +0100 (0:00:00.139)       0:00:30.537 ****** 
13:46:11 included: /var/lib/jenkins/jobs/deploy/workspace/roles/roleB/tasks/tasks.yml for www1, www2
13:46:11 
````

##### STEPS TO REPRODUCE

```yaml

```

##### EXPECTED RESULTS

````
$ ansible-playbook -i inv deploy.yml  -u ubuntu | sed "s/.insighthub.io//g"
[DEPRECATED] [defaults]hostfile: The key is misleading as it can also be a list of hosts, a directory or a list of paths. It will be removed in 2.8. As alternative use one of [inventory]
[DEPRECATION WARNING]: [defaults]hostfile option, The key is misleading as it can also be a list of hosts, a directory or a list of paths . This feature will be removed in version 2.8. Deprecation warnings can be disabled by setting
deprecation_warnings=False in ansible.cfg.

PLAY [Deploy ROLEA] **********************************************************************************************************************************************************************************************************************

TASK [ROLEA : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.035)       0:00:00.035 ******
ok: [www1] => {}

MSG:

We are in role: ROLEA, file main.yml


TASK [ROLEA : include_tasks] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.059)       0:00:00.094 ******
included: /var/lib/jenkins/ansible-tests/roles/ROLEA/tasks/tasks.yml for www1

TASK [ROLEA : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.031)       0:00:00.125 ******
ok: [www1] => {}

MSG:

We are in role: ROLEA, file main.yml


TASK [ROLEA : include roleB] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.059)       0:00:00.185 ******

TASK [ROLEC : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.040)       0:00:00.226 ******
ok: [www1] => {}

MSG:

We are in role: ROLEC, file main.yml


TASK [ROLEC : include_tasks] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.061)       0:00:00.287 ******
included: /var/lib/jenkins/ansible-tests/roles/ROLEC/tasks/tasks.yml for www1

TASK [ROLEC : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.031)       0:00:00.318 ******
ok: [www1] => {}

MSG:

We are in role: ROLEC
[include_role_bug.zip](https://github.com/ansible/ansible/files/1292727/include_role_bug.zip)

, file tasks.yml


TASK [ROLEB : include_tasks] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.061)       0:00:00.380 ******
included: /var/lib/jenkins/ansible-tests/roles/ROLEB/tasks/tasks.yml for www1

TASK [ROLEB : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.031)       0:00:00.412 ******
ok: [www1] => {}

MSG:

We are in role: ROLEB, file tasks.yml


PLAY RECAP *******************************************************************************************************************************************************************************************************************************
www1         : ok=8    changed=0    unreachable=0    failed=0

$

````
##### ACTUAL RESULTS

````
$ ansible-playbook -i inv deploy.yml  -u ubuntu | sed "s/.insighthub.io//g"
[DEPRECATED] [defaults]hostfile: The key is misleading as it can also be a list of hosts, a directory or a list of paths. It will be removed in 2.8. As alternative use one of [inventory]
[DEPRECATION WARNING]: [defaults]hostfile option, The key is misleading as it can also be a list of hosts, a directory or a list of paths . This feature will be removed in version 2.8. Deprecation warnings can be disabled by setting
deprecation_warnings=False in ansible.cfg.

PLAY [Deploy ROLEA] **********************************************************************************************************************************************************************************************************************

TASK [ROLEA : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.035)       0:00:00.035 ******
ok: [www1] => {}

MSG:

We are in role: ROLEA, file main.yml


TASK [ROLEA : include_tasks] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.059)       0:00:00.094 ******
included: /var/lib/jenkins/ansible-tests/roles/ROLEA/tasks/tasks.yml for www1

TASK [ROLEA : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.031)       0:00:00.125 ******
ok: [www1] => {}

MSG:

We are in role: ROLEA, file main.yml


TASK [ROLEA : include roleB] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.059)       0:00:00.185 ******

TASK [ROLEC : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.040)       0:00:00.226 ******
ok: [www1] => {}

MSG:

We are in role: ROLEC, file main.yml


TASK [ROLEC : include_tasks] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.061)       0:00:00.287 ******
included: /var/lib/jenkins/ansible-tests/roles/ROLEB/tasks/tasks.yml for www1

TASK [ROLEC : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.031)       0:00:00.318 ******
ok: [www1] => {}

MSG:

We are in role: ROLEB, file tasks.yml


TASK [ROLEB : include_tasks] *************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.061)       0:00:00.380 ******
included: /var/lib/jenkins/ansible-tests/roles/ROLEB/tasks/tasks.yml for www1

TASK [ROLEB : debug] *********************************************************************************************************************************************************************************************************************
Monday 11 September 2017  14:36:53 +0100 (0:00:00.031)       0:00:00.412 ******
ok: [www1] => {}

MSG:

We are in role: ROLEB, file tasks.yml


PLAY RECAP *******************************************************************************************************************************************************************************************************************************
www1         : ok=8    changed=0    unreachable=0    failed=0

Monday 11 September 2017  14:36:53 +0100 (0:00:00.060)       0:00:00.472 ******
===============================================================================
ROLEC : debug --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.06s
ROLEC : debug --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.06s
ROLEB : debug --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.06s
ROLEA : debug --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.06s
ROLEA : debug --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.06s
ROLEA : include roleB ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.04s
ROLEB : include_tasks ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.03s
ROLEC : include_tasks ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.03s
ROLEA : include_tasks ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.03s
````

```
ansible-playbook -i inv deploy.yml  -u ubuntu

```

**Comment by kraduk** [comment]
Here is the sample code

[include_role_bug.zip](https://github.com/ansible/ansible/files/1292737/include_role_bug.zip)

**Comment by alikins** [comment]
ping @bcoca

**Comment by programmeroftheeve** [comment]
This also affects_2.4

**Comment by duck-rh** [comment]
Another test case showing the difference of behavior between `roles` and `include_role`.

playbooks/z.yml:
```
---

- hosts: all
  # this works
  roles:
    - outer
  tags: case1

- hosts: all
  # this does not
  tasks:
    - include_role:
        name: outer
  tags: case2
```
roles/outer/tasks/main.yml:
```
---

- debug:
    msg: "in outer role"
```
roles/outer/meta/main.yml:
```
---
dependencies:
  - test
```
roles/test/tasks/main.yml:
```
---

- include_tasks: inc.yml
```
roles/test/tasks/inc.yml:
```
---

- debug:
    msg: "yatta!"
```

Result:
```
PLAY [all] ****************************************************************************************************************************************

TASK [test : include_tasks] ***********************************************************************************************************************
included: /home/duck/OSAS/Ansible/ansible_bug_nested_include_tasks/roles/test/tasks/inc.yml for localhost

TASK [test : debug] *******************************************************************************************************************************
ok: [localhost] => {
    "msg": "yatta!"
}

TASK [outer : debug] ******************************************************************************************************************************
ok: [localhost] => {
    "msg": "in outer role"
}

PLAY [all] ****************************************************************************************************************************************

TASK [test : include_tasks] ***********************************************************************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "reason": "Unable to retrieve file contents\nCould not find or access '/home/duck/OSAS/Ansible/ansible_bug_nested_include_tasks/playbooks/inc.yml'"}
	to retry, use: --limit @/home/duck/OSAS/Ansible/ansible_bug_nested_include_tasks/playbooks/z.retry

PLAY RECAP ****************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1   
```

**Comment by rlees85** [comment]
Noticed this today on 2.4.0.0. Thought I was loosing my marbles and trying to Google the problem was equally as amusing. Glad its been picked up already through!

**Comment by pilou-** [comment]
@duck-rh this one seems fixed by 9001a8794f19053f52a807cb59ced87357873e9e, could you check with test-case ?

**Comment by duck-rh** [comment]
Quack, works for me with the real-life case.

**Comment by pilou-** [comment]
@kraduk please, since this bug has been fixed, could you close this issue ?

**Comment by sivel** [comment]
This was resolved in 9001a8794f19053f52a807cb59ced87357873e9e and included in the 2.5.0 release.

Closing.
